src/request_builder.rs

```diff
@@ -0,0 +1,252 @@
+    /// # async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
+    /// let bytes = surf::get("https://httpbin.org/get").recv_bytes().await?;
+    /// assert!(bytes.len() > 0);
+    /// # Ok(()) }
+    /// ```
+    pub async fn recv_bytes(self) -> Result<Vec<u8>> {
+        let mut res = self.send().await?;
+        Ok(res.body_bytes().await?)
+    }
+
+    /// Submit the request and get the response body as a string.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # #[async_std::main]
+    /// # async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
+    /// let string = surf::get("https://httpbin.org/get").recv_string().await?;
+    /// assert!(string.len() > 0);
+    /// # Ok(()) }
+    /// ```
+    pub async fn recv_string(self) -> Result<String> {
+        let mut res = self.send().await?;
+        Ok(res.body_string().await?)
+    }
+
+    /// Submit the request and decode the response body from json into a struct.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # use serde::{Deserialize, Serialize};
+    /// # #[async_std::main]
+    /// # async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
+    /// #[derive(Deserialize, Serialize)]
+    /// struct Ip {
+    ///     ip: String
+    /// }
+    ///
+    /// let uri = "https://api.ipify.org?format=json";
+    /// let Ip { ip } = surf::get(uri).recv_json().await?;
+    /// assert!(ip.len() > 10);
+    /// # Ok(()) }
+    /// ```
+    pub async fn recv_json<T: serde::de::DeserializeOwned>(self) -> Result<T> {
+        let mut res = self.send().await?;
+        Ok(res.body_json::<T>().await?)
+    }
+
+    /// Submit the request and decode the response body from form encoding into a struct.
+    ///
+    /// # Errors
+    ///
+    /// Any I/O error encountered while reading the body is immediately returned
+    /// as an `Err`.
+    ///
+    /// If the body cannot be interpreted as valid json for the target type `T`,
+    /// an `Err` is returned.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # use serde::{Deserialize, Serialize};
+    /// # #[async_std::main]
+    /// # async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
+    /// #[derive(Deserialize, Serialize)]
+    /// struct Body {
+    ///     apples: u32
+    /// }
+    ///
+    /// let url = "https://api.example.com/v1/response";
+    /// let Body { apples } = surf::get(url).recv_form().await?;
+    /// # Ok(()) }
+    /// ```
+    pub async fn recv_form<T: serde::de::DeserializeOwned>(self) -> Result<T> {
+        let mut res = self.send().await?;
+        Ok(res.body_form::<T>().await?)
+    }
+
+    /// Return the constructed `Request`.
+    pub fn build(self) -> Request {
+        self.req.unwrap()
+    }
+
+    /// Create a `Client` and send the constructed `Request` from it.
+    pub fn send(self) -> BoxFuture<'static, Result<Response>> {
```

Comment:
Having either `send_with_client(self, client: &Client) -> BoxFuture<'static, Result<Response>>` or `with_client(self, &Client) -> Self` (storing the client and using it for `send`) would make the client reuse much nicer.

Reply:
```rust
let client = surf::client();
let res1 = client.send(surf::post(url).header().body()).await?;
let res2 = client.send(surf::post(url).header().body()).await?;
```

```rust
let client = surf::client();
let res1 = surf::post(url).header().body().with_client(&client).await?;
let res1 = surf::post(url).header().body().with_client(&client).await?;
```

I dunno, not only is the bottom longer but it's also weirder, injecting a reusable struct rather than having the reusable struct do the thing...? Doesn't feel right to me...